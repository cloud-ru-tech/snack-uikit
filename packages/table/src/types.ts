import {
  CellContext,
  ColumnDef,
  ColumnMeta,
  HeaderContext,
  PaginationState,
  RowSelectionOptions,
  RowSelectionState,
  SortingState,
} from '@tanstack/react-table';

import { ToolbarProps } from '@snack-ui/toolbar';

import { TableEmptyStateProps } from './components/TableEmptyState';
import { ColumnAlign, ColumnPinPosition } from './constants';
import { Except } from './helperComponents/types';

type BaseColumnDefinition<TData> = Except<
  ColumnDef<TData>,
  | 'footer'
  | 'enablePinning'
  | 'enableGrouping'
  | 'enableResizing'
  | 'enableColumnFilter'
  | 'filterFn'
  | 'enableGlobalFilter'
  | 'enableMultiSort'
  | 'enableHiding'
> & {
  /** Заголовок колонки */
  header?: string | ((ctx: HeaderContext<TData, unknown>) => string);
  /** Позиционирование контента ячейки в теле таблицы */
  align?: ColumnAlign;
  /** Отключить паддинг у ячейки в теле таблицы */
  noBodyCellPadding?: boolean;
  /** Отключить паддинг у ячейки в шапке таблицы */
  noHeaderCellPadding?: boolean;
  /** Отключить смещение бордера слева у ячейки хедера */
  noHeaderCellBorderOffset?: boolean;
  /** CSS-класс для ячейки в теле таблицы */
  cellClassName?: string;
  /** CSS-класс для ячейки в теле таблицы */
  headerClassName?: string;
  /** Пропуск в экспорте */
  meta?: ColumnMeta<TData, unknown> & { skipOnExport?: boolean };
};

type NormalColumnDefinition<TData> = BaseColumnDefinition<TData> & {
  pinned?: never;
};

type PinnedColumnDefinition<TData> = BaseColumnDefinition<TData> & {
  id: string;
  /** Закрепление колонки слева/справа в таблице */
  pinned: ColumnPinPosition;
  /** Размер ячейки */
  size: number;
};

export type ColumnDefinition<TData> = NormalColumnDefinition<TData> | PinnedColumnDefinition<TData>;

export type {
  RowActionInfo,
  RowActionProps,
  RowActionsColumnDefProps,
  StatusColumnDefinitionProps,
  RowInfo,
  RowClickHandler,
} from './helperComponents';

export {
  type PaginationState,
  type SortingState,
  type RowSelectionState,
  type RowSelectionOptions,
  type TableEmptyStateProps,
  type ToolbarProps,
  type HeaderContext,
  type CellContext,
};
