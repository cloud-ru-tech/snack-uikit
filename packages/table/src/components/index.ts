export * from './Table';
export * from './ServerTable';
export * from './types';
