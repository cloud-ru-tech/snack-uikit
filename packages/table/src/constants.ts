export const COLUMN_PIN_POSITION = {
  Left: 'left',
  Right: 'right',
} as const;

export const COLUMN_ALIGN = {
  Left: 'left',
  Right: 'right',
} as const;

export const COLUMN_SETTINGS_MODE = {
  /* Скрыто в настройках, значение всегда True */
  Hidden: 'hidden',
  DefaultTrue: 'defaultTrue',
  DefaultFalse: 'defaultFalse',
} as const;

export const TEST_IDS = {
  headerSortIndicator: 'table__header__sort-indicator',
  headerRow: 'table__header-row',
  headerCell: 'table__header-cell',
  bodyRow: 'table__body-row',
  bodyCell: 'table__body-cell',
  pinnedCells: 'table__pinned-cells',
  rowSelect: 'table__row-select',
  tree: {
    node: 'tree__node',
    chevron: 'tree__chevron',
    checkbox: 'tree__checkbox',
    radio: 'tree__radio',
    icon: 'tree__icon',
  },
  rowActions: {
    droplistTrigger: 'table__body-row__droplistTrigger',
    droplist: 'table__body-row__actions-droplist',
    option: 'list__base-item-option',
  },
  statusIndicator: 'table__status-indicator',
  statusLabel: 'table__status-label',
  toolbar: 'table__toolbar',
};

export const SORT_FN = {
  DateTime: 'datetime',
  AlphaNumeric: 'alphanumeric',
} as const;

export const DEFAULT_PAGE_SIZE = 10;
export const DEFAULT_SORTING = [];
export const DEFAULT_FILTER_VISIBILITY = [];
export const DEFAULT_ROW_SELECTION = {};

export enum DefaultColumns {
  Status = 'snack_predefined_statusColumn',
  Selection = 'selectionCell',
  RowActions = 'rowActions',
}
export const DEFAULT_COLUMNS = [DefaultColumns.Status, DefaultColumns.Selection, DefaultColumns.RowActions];
