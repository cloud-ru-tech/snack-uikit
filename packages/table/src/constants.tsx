export const COLUMN_PIN_POSITION = {
  Left: 'left',
  Right: 'right',
} as const;

export const COLUMN_ALIGN = {
  Left: 'left',
  Right: 'right',
} as const;

export const TEST_IDS = {
  headerSortIndicator: 'table__header__sort-indicator',
  headerRow: 'table__header-row',
  headerCell: 'table__header-cell',
  bodyRow: 'table__body-row',
  bodyCell: 'table__body-cell',
  pinnedCells: 'table__pinned-cells',
  rowSelect: 'table__row-select',
  tree: {
    node: 'tree__node',
    chevron: 'tree__chevron',
    checkbox: 'tree__checkbox',
    radio: 'tree__radio',
    icon: 'tree__icon',
  },
  rowActions: {
    droplistTrigger: 'table__body-row__droplistTrigger',
    droplist: 'table__body-row__actions-droplist',
    option: 'list__base-item-option',
  },
  statusIndicator: 'table__status-indicator',
  statusLabel: 'table__status-label',
  toolbar: 'table__toolbar',
};

export const SORT_FN = {
  DateTime: 'datetime',
  AlphaNumeric: 'alphanumeric',
} as const;

export const DEFAULT_PAGE_SIZE = 10;
