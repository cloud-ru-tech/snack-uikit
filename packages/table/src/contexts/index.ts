export * from './CellAutoResizeContext';
export { setNonce } from '@snack-uikit/scroll';
