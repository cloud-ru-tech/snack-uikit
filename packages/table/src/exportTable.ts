import { utils as xlsxUtils, writeFileXLSX } from 'xlsx';

import { ColumnDefinition } from './types';

type ExportTableData<TData> = {
  data: TData[];
  columnDefinitions: ColumnDefinition<TData>[];
  fileName?: string;
};

type ColumnKey<TData> = keyof ColumnDefinition<TData>;

function getColumnValue<TData>(column: ColumnDefinition<TData>, key: ColumnKey<TData>): string | undefined {
  if (key in column && column[key] && typeof column[key] !== 'object' && typeof column[key] !== 'function') {
    return String(column[key]);
  }
}

function getFilteredColumnsIds<TData extends object>(columnDefinitions: ColumnDefinition<TData>[]) {
  const accessorKey = 'accessorKey' as ColumnKey<TData>;

  return (columnDefinitions as ColumnDefinition<TData>[])
    .filter(column => getColumnValue(column, accessorKey) && !column.meta?.skipOnExport)
    .map(column => getColumnValue(column, accessorKey));
}

function getFilteredColumnsHeaders<TData extends object>(columnDefinitions: ColumnDefinition<TData>[]) {
  const accessorKey = 'accessorKey' as ColumnKey<TData>;
  const header = 'header' as ColumnKey<TData>;

  return (columnDefinitions as ColumnDefinition<TData>[])
    .filter(column => getColumnValue(column, accessorKey) && !column.meta?.skipOnExport)
    .map(column => getColumnValue(column, header) || getColumnValue(column, accessorKey));
}

function getXlsxFormatTable<TData extends object>({
  data,
  columnDefinitions,
}: {
  data: TData[];
  columnDefinitions: ColumnDefinition<TData>[];
}) {
  const filteredIds = getFilteredColumnsIds(columnDefinitions);
  return data.map((line: TData) => {
    const lineRecord = line as Record<string, string>;
    const result: string[] = [];
    filteredIds.forEach(key => {
      if (!key) {
        result.push('');
        return;
      }

      result.push(lineRecord[key]);
    });
    return result;
  });
}

export function exportToCSV<TData extends object>({
  columnDefinitions,
  fileName = 'Table',
  data,
}: ExportTableData<TData>) {
  const xlsxData = getXlsxFormatTable({ data, columnDefinitions });
  const headers = getFilteredColumnsHeaders(columnDefinitions);
  const table = [headers, ...xlsxData];
  const csv = table.map(line => line.map(el => (el === undefined ? `""` : `"${el}"`)).join(',')).join('\n');

  const utf8Prefix = new Uint8Array([0xef, 0xbb, 0xbf]);
  const blob = new Blob([utf8Prefix, csv], { type: 'text/csv' });
  const url = window.URL.createObjectURL(blob);
  const tempLink = Object.assign(document.createElement('a'), {
    target: '_blank',
    href: url,
    download: fileName,
  });
  tempLink.click();
  tempLink.remove();

  return csv;
}

export function exportToXLSX<TData extends object>({
  columnDefinitions,
  fileName = 'Table',
  data,
}: ExportTableData<TData>) {
  const xlsxData = getXlsxFormatTable({ data, columnDefinitions });
  const headers = getFilteredColumnsHeaders(columnDefinitions);
  const workbook = xlsxUtils.book_new();
  const worksheet = xlsxUtils.aoa_to_sheet([headers, ...xlsxData]);
  xlsxUtils.book_append_sheet(workbook, worksheet);
  writeFileXLSX(workbook, `${fileName}.xlsx`);

  return worksheet;
}
