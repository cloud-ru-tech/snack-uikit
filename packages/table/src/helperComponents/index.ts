export * from './Cells';
export * from './Rows';
export * from './ExportButton';
export * from './contexts';
export * from './helpers';
export * from './hooks';
export * from './types';
export * from './TableEmptyState';
export * from './TablePagination';
export * from './ColumnsSettings';
