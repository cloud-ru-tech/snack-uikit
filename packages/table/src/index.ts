export * from './components';
export * from './types';
export * from './exportTable';

export { type CopyCellProps, type ActionsGenerator, CopyCell } from './helperComponents';
