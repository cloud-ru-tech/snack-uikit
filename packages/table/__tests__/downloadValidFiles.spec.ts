import { vi } from 'vitest';

vi.mock('xlsx', async () => ({
  ...(await vi.importActual<object>('xlsx')),
  writeFileXLSX: vi.fn(),
}));

import { utils as xlsxUtils, writeFileXLSX } from 'xlsx';

import { ColumnDefinition } from '../src';
import { exportToXLSX } from '../src/exportTable';

type StubData = {
  status: string;
  col1: number;
  col2: number;
  col3: number;
  col4: number;
  date: number;
};

describe('downloadValidFiles', () => {
  afterAll(() => {
    vi.clearAllMocks();
  });

  const columnDefinitions: ColumnDefinition<StubData>[] = [
    { id: '1', accessorKey: 'col1', size: 140, enableSorting: true, sortDescFirst: true },
    { id: '2', accessorKey: 'col2', size: 140, enableSorting: true },
    { id: '3', accessorKey: 'col3', minSize: 110 },
    { id: '4', accessorKey: 'col4', enableSorting: true },
    { id: '5', accessorKey: 'date', enableSorting: true, size: 146 },
  ];

  const mockData = [
    { status: 'Loading', col2: 0, col4: 0, col1: 0, col3: 0, date: 1337919743656 },
    { status: 'Green', col3: 1, col1: 1, col4: 1, col2: 1, date: 1285032902056 },
  ];
  const fileName = 'TestTable';

  it('download XLSX files with elements in the correct order', () => {
    const expectedResult = [
      {
        col1: 0,
        col2: 0,
        col3: 0,
        col4: 0,
        date: 1337919743656,
      },
      {
        col1: 1,
        col2: 1,
        col3: 1,
        col4: 1,
        date: 1285032902056,
      },
    ];

    const result = xlsxUtils.sheet_to_json(exportToXLSX({ columnDefinitions, data: mockData, fileName }));

    expect(result).toEqual(expectedResult);

    expect(writeFileXLSX).toBeCalledTimes(1);

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    expect(writeFileXLSX.mock.calls[0]).toMatchSnapshot();
  });
});
