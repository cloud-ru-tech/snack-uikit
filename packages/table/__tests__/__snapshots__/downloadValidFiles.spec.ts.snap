// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`downloadValidFiles > download XLSX files with elements in the correct order 1`] = `
[
  {
    "SheetNames": [
      "Sheet1",
    ],
    "Sheets": {
      "Sheet1": {
        "!ref": "A1:E3",
        "A1": {
          "t": "s",
          "v": "col1",
          "w": "col1",
        },
        "A2": {
          "t": "n",
          "v": 0,
        },
        "A3": {
          "t": "n",
          "v": 1,
        },
        "B1": {
          "t": "s",
          "v": "col2",
          "w": "col2",
        },
        "B2": {
          "t": "n",
          "v": 0,
        },
        "B3": {
          "t": "n",
          "v": 1,
        },
        "C1": {
          "t": "s",
          "v": "col3",
          "w": "col3",
        },
        "C2": {
          "t": "n",
          "v": 0,
        },
        "C3": {
          "t": "n",
          "v": 1,
        },
        "D1": {
          "t": "s",
          "v": "col4",
          "w": "col4",
        },
        "D2": {
          "t": "n",
          "v": 0,
        },
        "D3": {
          "t": "n",
          "v": 1,
        },
        "E1": {
          "t": "s",
          "v": "date",
          "w": "date",
        },
        "E2": {
          "t": "n",
          "v": 1337919743656,
        },
        "E3": {
          "t": "n",
          "v": 1285032902056,
        },
      },
    },
  },
  "TestTable.xlsx",
]
`;
